agent_app_sample_code/agents/function_calling_agent_w_retriever_tool.py

```diff
@@ -0,0 +1,496 @@
+                        page_content=page_content, metadata=metadata, type="Document"
+                    )  # , 9)
+                    docs.append(doc)
+
+        return docs
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ##### Agent
+
+# COMMAND ----------
+
+class AgentWithRetriever(mlflow.pyfunc.PythonModel):
+    """
+    Class representing an Agent that includes a Retriever tool
+    """
+
+    def __init__(self):
+        self.config = mlflow.models.ModelConfig(
+            development_config="agents/generated_configs/agent.yaml"
+        )
+
+        # Load the LLM
+        self.model_serving_client = get_deploy_client("databricks")
+
+        # Init the Retriever tool
+        self.retriever_tool = VectorSearchRetriever(self.config.get("retriever_tool"))
+
+        # Configure the Review App to use the Retriever's schema
+        vector_search_schema = self.config.get("retriever_tool").get(
+            "vector_search_schema"
+        )
+        mlflow.models.set_retriever_schema(
+            primary_key=vector_search_schema.get("primary_key"),
+            text_column=vector_search_schema.get("chunk_text"),
+            doc_uri=vector_search_schema.get("doc_uri"),
+        )
+
+        self.tool_functions = {
+            "retrieve_documents": self.retriever_tool,
+        }
+
+        # Internal representation of the chat history.  As the Agent iteratively selects/executes tools, the history will be stored here.  Since the Agent is stateless, this variable must be populated on each invocation of `predict(...)`.
+        self.chat_history = None
+
+    @mlflow.trace(name="chain", span_type="CHAIN")
+    def predict(
+        self,
+        context: Any = None,
+        model_input: Union[ChatCompletionRequest, Dict, pd.DataFrame] = None,
+        params: Any = None,
+    ) -> StringResponse:
+        ##############################################################################
+        # Extract `messages` key from the `model_input`
+        messages = self.get_messages_array(model_input)
+
+        ##############################################################################
+        # Parse `messages` array into the user's query & the chat history
+        with mlflow.start_span(name="parse_input", span_type="PARSER") as span:
+            span.set_inputs({"messages": messages})
+            user_query = self.extract_user_query_string(messages)
+            # Save the history inside the Agent's internal state
+            self.chat_history = self.extract_chat_history(messages)
+            span.set_outputs(
+                {"user_query": user_query, "chat_history": self.chat_history}
+            )
+
+        ##############################################################################
+        # Generate Answer
+        system_prompt = self.config.get("llm_config").get("llm_system_prompt_template")
+
+        # Add the previous history
+        # TODO: Need a way to include the previous tool calls
```

Comment:
Can we address this TODO?

Reply:
Specifically is there a simple way to update our output logic to include tool output info? Looks like `recursively_call_and_run_tools` already adds tool messages to some list of messages

Reply:
i believe this is already taken care of actually. the tool call output is returned to the client and re-sent in with the next requst